docs/madr/decisions/057-rework-mesh-routes-matching.md

```diff
@@ -0,0 +1,533 @@
+# Rework of Mesh*Routes matching
+
+* Status: accepted
+
+Technical story: https://github.com/kumahq/kuma/issues/10247
+
+## Context and Problem Statement
+
+At the moment when targeting `Mesh*Route` you need to put it in topLevel targetRef. 
+We want to iterate over policy design to fully utilize the fact that we are referencing real objects.
+
+## Decision Outcome
+
+- add possibility to put `Mesh*Routes` in `spec.to[].targetRef`
+
+## Considered options
+
+- add possibility to put `Mesh*Routes` in `spec.to[].targetRef`
+- leave `Mesh*Routes` in top level targetRef as it is now
+
+### Add possibility to put Mesh*Routes in `spec.to[].targetRef`
+
+#### Advantages
+
+- simplifying policy api as `spec.targetRef` will always pick proxy and `spec.to[].targetRef` section will pick outbounds
+- ability to set different confs on different proxy for the same route (overriding default conf for route without producer/consumer policies)
+- no need for specifying `spec.targetRef` when targeting `Mesh`
+- easier policy matching code
+
+#### Disadvantages
+
+- can be harder to understand because `spec.targetRef` can be MeshSubset 
```

Comment:
One suggestion here was to make `spec.targetRef` optional so most people wouldn't even know it exists

Reply:
And it is treated as optional in next sections of MADR, I've updated this point to make it clear that this would be rather rare case

Reply:
> One suggestion here was to make spec.targetRef optional so most people wouldn't even know it exists

This is kind of the problem I have with the whole thing. We're basically admitting `spec.targetRef` is almost always going to be empty. So why does it have the top level place? Why not just have `spec.workloadTargetRef` that's optional and keep the most important info at `spec.targetRef`?

Reply:
so you want `spec.workloadTargetRef` to select entire proxies, like now `spec.targetRef` does in this madr?

But what about MeshService? should this be still in `spec.to[].targetRef` or as route in `spec.targetRef`? What is the difference here?

Reply:
In my ideal world it "should" be in `spec.targetRef`. The `spec.workloadTargetRef` is also 99% of the time going to be either Mesh or namespace. The most interesting part of the policy is the outbound it affects.

Reply:
Got it, I will wrap this up and present this as another option to consider, also I will try to present some use cases with examples

Reply:
>  it doesn't change the way MeshHTTPRoute is used

it would make this MADR unnecessary right? MeshHTTPRoute is already in top level

> just syntactical changes

so this MADR is also just syntactical changes then

Reply:
The change in this MADR is not syntactical. We change the semantic of `Policy <-> MeshHTTPRoute` reference from:
```
Policy targets MeshHTTPRoute and because of that targets all the places when the route is applied.
```
to
```
Policy targets the place where MeshHTTPRoute is applied (same route can have different confs on different DPPs).
```

> I don't think it's syntactical. This is about making the semantics of spec.targetRef and by extension API the most useful and user friendly it can be

I don't know, you can kind of convert one to another by using a template:
```yaml
targetRef: ${{ policy.to[0].targetRef }}
conf: ${{ policy.to[0].conf }}
workloadTargetRef: ${{ policy.targetRef }}
```

Reply:
First of all I think you're forgetting about what would happen to `from`.

> Policy targets MeshHTTPRoute and because of that targets all the places when the route is applied.
> Policy targets the place where MeshHTTPRoute is applied (same route can have different confs on different DPPs).

I think this is just playing with "targeting". Let's be very honest. We're always "targeting" two things. That's why we have two `targetRefs`.

We're always "targeting the place where MeshHTTPRoute is applied" because we're targeting the MeshHTTPRoute, it's just in `to`.

You can also mechanically translate the two options in this MADR, can't you?
`spec.to[0].targetRef: ${{ spec.targetRef }}`

All you're doing is essentially adding the ability to define `workloadTargetRef`, except it's called `targetRef`.

Anyway I don't think any of that matters either way. We're playing around with the structure of policies already. We're literally defining the API. I am unconvinced that this change actually improves anything when it comes to UX.

Reply:
Put a different way, I think it would be more of a UX improvement if we put important things in `spec.targetRef` everywhere than if we put MeshHTTPRoute in `spec.to[].targetRef`

Reply:
I see

```
## Context and Problem Statement

At the moment when targeting `Mesh*Route` you need to put it in topLevel targetRef. 
We want to iterate over policy design to fully utilize the fact that we are referencing real objects.
```

We're already referencing real objects as far as I can tell?
What is the actual problem we're solving?

Reply:
this has been integrated into the MADR